go/libraries/doltcore/sqle/read_replica_database.go

```diff
@@ -131,132 +151,189 @@ func (rrd ReadReplicaDatabase) PullFromRemote(ctx *sql.Context) error {
+				return nil
+			}
+		}
+
+		remoteRefs = prunedRefs
+		err = pullBranches(ctx, rrd, remoteRefs, localRefs, currentBranchRef, behavior)
+
+		if err != nil && !SkipReplicationWarnings() {
 			return err
+		} else if err != nil {
+			ctx.GetLogger().Warn(err.Error())
+			return nil
 		}
-		toPull, toDelete, err := getReplicationBranches(ctx, rrd)
-		err = pullBranches(ctx, rrd, toPull, currentBranchRef, behavior)
-		if err != nil {
+
+	case allHeads == int8(1):
+		err = pullBranches(ctx, rrd, remoteRefs, localRefs, currentBranchRef, behavior)
+		if err != nil && !SkipReplicationWarnings() {
 			return err
+		} else if err != nil {
+			ctx.GetLogger().Warn(err.Error())
+			return nil
 		}
-		err = deleteBranches(ctx, rrd, toDelete, currentBranchRef)
-		if err != nil {
+
+		err = deleteBranches(ctx, rrd, toDelete)
+		if err != nil && !SkipReplicationWarnings() {
 			return err
+		} else if err != nil {
+			ctx.GetLogger().Warn(err.Error())
+			return nil
 		}
 	default:
 		return fmt.Errorf("%w: dolt_replicate_heads not set", ErrInvalidReplicateHeadsSetting)
 	}
+
 	return nil
 }
 
+func (rrd ReadReplicaDatabase) RebaseSourceDb(ctx *sql.Context) error {
+	return rrd.srcDB.Rebase(ctx)
+}
+
 type pullBehavior bool
 
 const pullBehavior_fastForward pullBehavior = false
 const pullBehavior_forcePull pullBehavior = true
 
-func pullBranches(ctx *sql.Context, rrd ReadReplicaDatabase, branches []string, currentBranchRef ref.DoltRef, behavior pullBehavior) error {
-	refSpecs, err := env.ParseRSFromArgs(rrd.remote.Name, branches)
-	if err != nil {
-		return err
+// pullBranches pulls the remote branches named. If a corresponding local branch exists, it will be fast-forwarded. If
+// it doesn't exist, it will be created.
+func pullBranches(
+	ctx *sql.Context,
+	rrd ReadReplicaDatabase,
+	remoteRefs []doltdb.RefWithHash,
+	localRefs []doltdb.RefWithHash,
+	currentBranchRef ref.DoltRef,
+	behavior pullBehavior,
+) error {
+	localHashesByRef := make(map[string]doltdb.RefWithHash)
+	remoteHashesByRef := make(map[string]doltdb.RefWithHash)
+
+	for _, b := range remoteRefs {
+		remoteHashesByRef[b.Ref.GetPath()] = b
 	}
 
-	for i, refSpec := range refSpecs {
-		branch := ref.NewBranchRef(branches[i])
-		rtRef := refSpec.DestRef(branch)
+	for _, b := range localRefs {
+		localHashesByRef[b.Ref.GetPath()] = b
+	}
 
-		fetchCtx := ctx.Context
+	_, err := rrd.limiter.Run(ctx, "-all", func() (any, error) {
+		srcRoot, err := rrd.srcDB.NomsRoot(ctx)
+		if err != nil {
+			return nil, err
+		}
 
-		srcDBCommitA, err := rrd.limiter.Run(ctx, branch.String(), func() (any, error) {
-			// fetch remote tracking ref only
-			var srcDBCommit *doltdb.Commit
-			{
-				srcDBCommit, err = actions.FetchRemoteBranch(fetchCtx, rrd.tmpDir, rrd.remote, rrd.srcDB, rrd.ddb, branch, actions.NoopRunProgFuncs, actions.NoopStopProgFuncs)
-				if err != nil {
-					return nil, err
+		err = rrd.ddb.PullChunks(ctx, rrd.tmpDir, rrd.srcDB, srcRoot, nil, nil)
```

Comment:
Nit: It seems more correct, clearer and very marginally more performance to make an interface that will pull more than one addr at once and then to pass all the addrs in `remoteRefs` to it.